File: backend/app/api/v1/hierarchy.py
"""层级结构 API

提供 GB50300 层级结构的查询接口
"""

from typing import Optional
from fastapi import APIRouter, HTTPException, status, Query, Depends

from app.services.hierarchy import HierarchyService
from app.models.api.hierarchy import (
    ProjectListItem,
    ProjectDetail,
    HierarchyResponse,
    BuildingDetail,
    DivisionDetail,
    SubDivisionDetail,
    ItemDetail,
    InspectionLotDetail,
)
from app.utils.memgraph import get_memgraph_client, MemgraphClient


router = APIRouter(prefix="/hierarchy", tags=["hierarchy"])


def get_hierarchy_service(
    client: MemgraphClient = Depends(get_memgraph_client)
) -> HierarchyService:
    """获取 Hierarchy Service 实例（依赖注入）"""
    return HierarchyService(client=client)


@router.get(
    "/projects",
    response_model=dict,
    summary="获取项目列表",
    description="获取所有项目的列表，支持分页"
)
async def get_projects(
    page: int = Query(default=1, ge=1, description="页码"),
    page_size: int = Query(default=20, ge=1, le=100, description="每页数量"),
    service: HierarchyService = Depends(get_hierarchy_service),
) -> dict:
    """获取项目列表"""
    result = service.get_projects(page=page, page_size=page_size)
    return {
        "status": "success",
        "data": result,
    }


@router.get(
    "/projects/{project_id}",
    response_model=dict,
    summary="获取项目详情",
    description="根据项目 ID 获取项目详细信息"
)
async def get_project(
    project_id: str,
    service: HierarchyService = Depends(get_hierarchy_service),
) -> dict:
    """获取项目详情"""
    project = service.get_project_detail(project_id)
    
    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Project not found: {project_id}",
        )
    
    return {
        "status": "success",
        "data": project.model_dump(),
    }


@router.get(
    "/projects/{project_id}/hierarchy",
    response_model=dict,
    summary="获取项目层级树",
    description="获取项目的完整层级结构树"
)
async def get_project_hierarchy(
    project_id: str,
    service: HierarchyService = Depends(get_hierarchy_service),
) -> dict:
    """获取项目层级树"""
    hierarchy = service.get_project_hierarchy(project_id)
    
    if not hierarchy:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Project not found: {project_id}",
        )
    
    return {
        "status": "success",
        "data": hierarchy.model_dump(),
    }


@router.get(
    "/buildings/{building_id}",
    response_model=dict,
    summary="获取单体详情",
    description="根据单体 ID 获取单体详细信息"
)
async def get_building(
    building_id: str,
    service: HierarchyService = Depends(get_hierarchy_service),
) -> dict:
    """获取单体详情"""
    building = service.get_building_detail(building_id)
    
    if not building:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Building not found: {building_id}",
        )
    
    return {
        "status": "success",
        "data": building.model_dump(),
    }


@router.get(
    "/divisions/{division_id}",
    response_model=dict,
    summary="获取分部详情",
    description="根据分部 ID 获取分部详细信息"
)
async def get_division(
    division_id: str,
    service: HierarchyService = Depends(get_hierarchy_service),
) -> dict:
    """获取分部详情"""
    division = service.get_division_detail(division_id)
    
    if not division:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Division not found: {division_id}",
        )
    
    return {
        "status": "success",
        "data": division.model_dump(),
    }


@router.get(
    "/subdivisions/{subdivision_id}",
    response_model=dict,
    summary="获取子分部详情",
    description="根据子分部 ID 获取子分部详细信息"
)
async def get_subdivision(
    subdivision_id: str,
    service: HierarchyService = Depends(get_hierarchy_service),
) -> dict:
    """获取子分部详情"""
    subdivision = service.get_subdivision_detail(subdivision_id)
    
    if not subdivision:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"SubDivision not found: {subdivision_id}",
        )
    
    return {
        "status": "success",
        "data": subdivision.model_dump(),
    }


@router.get(
    "/items/{item_id}",
    response_model=dict,
    summary="获取分项详情",
    description="根据分项 ID 获取分项详细信息（包含检验批列表）"
)
async def get_item(
    item_id: str,
    service: HierarchyService = Depends(get_hierarchy_service),
) -> dict:
    """获取分项详情"""
    item = service.get_item_detail(item_id)
    
    if not item:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Item not found: {item_id}",
        )
    
    return {
        "status": "success",
        "data": item.model_dump(),
    }


@router.get(
    "/inspection-lots/{lot_id}",
    response_model=dict,
    summary="获取检验批详情",
    description="根据检验批 ID 获取检验批详细信息（包含构件列表）"
)
async def get_inspection_lot(
    lot_id: str,
    service: HierarchyService = Depends(get_hierarchy_service),
) -> dict:
    """获取检验批详情"""
    lot = service.get_inspection_lot_detail(lot_id)
    
    if not lot:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"InspectionLot not found: {lot_id}",
        )
    
    return {
        "status": "success",
        "data": lot.model_dump(),
    }

File: frontend/src/services/hierarchy.ts
/** 层级结构 API 服务 */

import { apiGet } from './api';
import { ApiResponse, PaginatedResponse } from './api';

/**
 * 项目列表项
 */
export interface ProjectListItem {
  id: string;
  name: string;
  description?: string;
  building_count?: number;
  created_at: string;
  updated_at: string;
}

/**
 * 项目详情
 */
export interface ProjectDetail {
  id: string;
  name: string;
  description?: string;
  building_count?: number;
  created_at: string;
  updated_at: string;
}

/**
 * 层级节点
 */
export interface HierarchyNode {
  id: string;
  label: string;
  name: string;
  children: HierarchyNode[];
  metadata?: Record<string, any>;
}

/**
 * 层级树响应
 */
export interface HierarchyResponse {
  project_id: string;
  project_name: string;
  hierarchy: HierarchyNode;
}

/**
 * 单体详情
 */
export interface BuildingDetail {
  id: string;
  name: string;
  project_id: string;
  description?: string;
  created_at: string;
  updated_at: string;
}

/**
 * 分部详情
 */
export interface DivisionDetail {
  id: string;
  name: string;
  building_id: string;
  description?: string;
  created_at: string;
  updated_at: string;
}

/**
 * 子分部详情
 */
export interface SubDivisionDetail {
  id: string;
  name: string;
  division_id: string;
  description?: string;
  created_at: string;
  updated_at: string;
}

/**
 * 分项详情
 */
export interface ItemDetail {
  id: string;
  name: string;
  subdivision_id: string;
  description?: string;
  inspection_lot_count?: number;
  created_at: string;
  updated_at: string;
}

/**
 * 检验批详情
 */
export interface InspectionLotDetail {
  id: string;
  name: string;
  item_id: string;
  spatial_scope?: string;
  status: 'PLANNING' | 'IN_PROGRESS' | 'SUBMITTED' | 'APPROVED' | 'PUBLISHED';
  element_count?: number;
  created_at: string;
  updated_at: string;
}

/**
 * 获取项目列表
 */
export async function getProjects(
  page: number = 1,
  pageSize: number = 20
): Promise<PaginatedResponse<ProjectListItem>> {
  const response = await apiGet<ApiResponse<PaginatedResponse<ProjectListItem>>>(
    `/api/v1/hierarchy/projects?page=${page}&page_size=${pageSize}`
  );
  return response.data;
}

/**
 * 获取项目详情
 */
export async function getProjectDetail(projectId: string): Promise<ProjectDetail> {
  const response = await apiGet<ApiResponse<ProjectDetail>>(
    `/api/v1/hierarchy/projects/${projectId}`
  );
  return response.data;
}

/**
 * 获取项目层级树
 */
export async function getProjectHierarchy(projectId: string): Promise<HierarchyResponse> {
  const response = await apiGet<ApiResponse<HierarchyResponse>>(
    `/api/v1/hierarchy/projects/${projectId}/hierarchy`
  );
  return response.data;
}

/**
 * 获取单体详情
 */
export async function getBuildingDetail(buildingId: string): Promise<BuildingDetail> {
  const response = await apiGet<ApiResponse<BuildingDetail>>(
    `/api/v1/hierarchy/buildings/${buildingId}`
  );
  return response.data;
}

/**
 * 获取分部详情
 */
export async function getDivisionDetail(divisionId: string): Promise<DivisionDetail> {
  const response = await apiGet<ApiResponse<DivisionDetail>>(
    `/api/v1/hierarchy/divisions/${divisionId}`
  );
  return response.data;
}

/**
 * 获取子分部详情
 */
export async function getSubDivisionDetail(subdivisionId: string): Promise<SubDivisionDetail> {
  const response = await apiGet<ApiResponse<SubDivisionDetail>>(
    `/api/v1/hierarchy/subdivisions/${subdivisionId}`
  );
  return response.data;
}

/**
 * 获取分项详情
 */
export async function getItemDetail(itemId: string): Promise<ItemDetail> {
  const response = await apiGet<ApiResponse<ItemDetail>>(
    `/api/v1/hierarchy/items/${itemId}`
  );
  return response.data;
}

/**
 * 获取检验批详情
 */
export async function getInspectionLotDetail(lotId: string): Promise<InspectionLotDetail> {
  const response = await apiGet<ApiResponse<InspectionLotDetail>>(
    `/api/v1/hierarchy/inspection-lots/${lotId}`
  );
  return response.data;
}

